File: psc-frontend/src/schemas/audit/finding.ts
import { z } from 'zod';

export const FINDING_TYPES = ['NC', 'OBSERVATION'] as const;
export const NC_CATEGORIES = ['MAJOR_NC', 'MINOR_NC'] as const;
export const OBSERVATION_CATEGORIES = ['OBSERVATION', 'IMPROVEMENT_SUGGESTION', 'OFI'] as const;
export const RULE_BOOK_TYPES = ['ISM', 'ISPS', 'MLC', 'SOLAS', 'STCW', 'MARPOL', 'COLREG', 'KSM_SMS', 'FLAG', 'OTHER'] as const;
export const FINDING_PRIORITIES = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL'] as const;
export const CERTIFICATE_IMPACTS = ['NONE', 'CERT_VALID', 'RENEWAL_AT_RISK', 'SUSPENDED', 'WITHDRAWN'] as const;

export interface AuditClauseMasterRow {
  id: string;
  code: string;
  title: string;
  code_version: string;
}

export interface AuditClauseMaster {
  rule_book_type: string;
  clauses: AuditClauseMasterRow[];
}

export const auditFindingClauseSchema = z.object({
  rule_book_type: z.enum(RULE_BOOK_TYPES),
  rule_clause_id: z.string().uuid().or(z.literal('')).optional().default(''),
  clause_ref_text: z.string().max(200).optional().default(''),
  clause_subref_text: z.string().max(200).optional().default(''),
  is_primary: z.boolean().default(false),
});

export const auditFindingCreateSchema = z
  .object({
    finding_type: z.enum(FINDING_TYPES),
    nc_category: z.enum(NC_CATEGORIES).or(z.literal('')).optional().default(''),
    observation_category: z.enum(OBSERVATION_CATEGORIES).or(z.literal('')).optional().default('OBSERVATION'),
    standard_code: z.string().max(20).optional().default('ISM'),
    description: z.string().min(1, 'Description is required'),
    objective_evidence: z.string().optional().default(''),
    def_code_id: z.string().min(1, 'DefCode is required').max(5),
    checklist_item_id: z.string().uuid().or(z.literal('')).optional().default(''),
    priority: z.enum(FINDING_PRIORITIES).optional().default('MEDIUM'),
    certificate_impact: z.enum(CERTIFICATE_IMPACTS).or(z.literal('')).optional().default(''),
    certificates_at_risk: z.string().max(100).optional().default(''),
    is_fleetwide_relevance: z.boolean().optional().default(false),
    clauses: z.array(auditFindingClauseSchema).min(1, 'At least one clause reference is required'),
  })
  .refine((data) => (data.finding_type === 'NC' ? Boolean(data.nc_category) : true), {
    message: 'NC category is required',
    path: ['nc_category'],
  })
  .refine((data) => (data.finding_type === 'OBSERVATION' ? Boolean(data.observation_category) : true), {
    message: 'Observation category is required',
    path: ['observation_category'],
  })
  .refine((data) => (data.finding_type === 'NC' ? true : !data.is_fleetwide_relevance), {
    message: 'Fleet-wide relevance is available only for NC findings',
    path: ['is_fleetwide_relevance'],
  })
  .refine((data) => data.clauses.filter((clause) => clause.is_primary).length === 1, {
    message: 'Exactly one clause reference must be primary',
    path: ['clauses'],
  })
  .refine(
    (data) =>
      data.clauses.every((clause) => {
        if (clause.rule_book_type === 'OTHER' || clause.rule_book_type === 'FLAG') {
          const length = clause.clause_ref_text.trim().length;
          return length >= 5 && length <= 200 && !clause.rule_clause_id;
        }
        return Boolean(clause.rule_clause_id);
      }),
    {
      message: 'Select a seeded clause, or enter 5-200 characters for OTHER/FLAG.',
      path: ['clauses'],
    }
  );

export type AuditFindingCreateFormData = z.infer<typeof auditFindingCreateSchema>;

export interface AuditFindingCreateResponse {
  id: string;
  audit_detail_id: string;
  finding_type: 'NC' | 'OBSERVATION' | string;
  nc_category: string | null;
  observation_category: string | null;
  standard_code: string | null;
  rule_book_type: string | null;
  rule_clause_id: string | null;
  clause_ref_text: string | null;
  description: string;
  objective_evidence: string | null;
  priority: string;
  certificates_at_risk: string | null;
  is_fleetwide_relevance: boolean;
  linked_circular_id: string | null;
  psc_deficiency_id: string;
  car_id: string;
  car_number: string;
  car_status: string;
  created: boolean;
}

export interface AuditIssueCircularResponse {
  status: string;
  circular_id: string;
  detail_url: string;
  payload: Record<string, unknown>;
}

export function findingDefaults(checklistItemId?: string): AuditFindingCreateFormData {
  return {
    finding_type: 'NC',
    nc_category: 'MINOR_NC',
    observation_category: 'OBSERVATION',
    standard_code: 'ISM',
    description: '',
    objective_evidence: '',
    def_code_id: '10101',
    checklist_item_id: checklistItemId || '',
    priority: 'MEDIUM',
    certificate_impact: '',
    certificates_at_risk: '',
    is_fleetwide_relevance: false,
    clauses: [
      {
        rule_book_type: 'ISM',
        rule_clause_id: '',
        clause_ref_text: '',
        clause_subref_text: '',
        is_primary: true,
      },
    ],
  };
}

File: psc-backend/apps/inspection/audit/serializers/finding.py
"""Serializers for Audit finding capture."""

from __future__ import annotations

from django.db import connection
from rest_framework import serializers

from apps.inspection.audit.models import AuditFindingClause
from apps.inspection.audit.services.finding import (
    FINDING_TYPES,
    NC_CATEGORIES,
    OBSERVATION_CATEGORIES,
    PRIORITIES,
    RULE_BOOK_TYPES,
)


CERTIFICATE_IMPACTS = {"NONE", "CERT_VALID", "RENEWAL_AT_RISK", "SUSPENDED", "WITHDRAWN"}


class AuditFindingClauseSerializer(serializers.Serializer):
    rule_book_type = serializers.ChoiceField(choices=tuple(sorted(RULE_BOOK_TYPES)))
    rule_clause_id = serializers.UUIDField(required=False, allow_null=True)
    clause_ref_text = serializers.CharField(required=False, allow_blank=True, max_length=200)
    clause_subref_text = serializers.CharField(required=False, allow_blank=True, max_length=200)
    is_primary = serializers.BooleanField(default=False)


class AuditFindingCreateSerializer(serializers.Serializer):
    finding_type = serializers.ChoiceField(choices=tuple(sorted(FINDING_TYPES)))
    nc_category = serializers.ChoiceField(choices=tuple(sorted(NC_CATEGORIES)), required=False, allow_blank=True)
    observation_category = serializers.ChoiceField(
        choices=tuple(sorted(OBSERVATION_CATEGORIES)),
        required=False,
        allow_blank=True,
    )
    standard_code = serializers.CharField(required=False, allow_blank=True, max_length=20)
    description = serializers.CharField()
    objective_evidence = serializers.CharField(required=False, allow_blank=True)
    def_code_id = serializers.CharField(required=False, allow_blank=True, max_length=5)
    def_code = serializers.CharField(required=False, allow_blank=True, max_length=10)
    checklist_item_id = serializers.UUIDField(required=False, allow_null=True)
    clauses = AuditFindingClauseSerializer(many=True)
    psc_deficiency_id = serializers.UUIDField(required=False, allow_null=True)
    priority = serializers.ChoiceField(choices=tuple(sorted(PRIORITIES)), required=False, allow_blank=True)
    original_due_date = serializers.DateField(required=False, allow_null=True)
    certificate_impact = serializers.ChoiceField(
        choices=tuple(sorted(CERTIFICATE_IMPACTS)),
        required=False,
        allow_blank=True,
    )
    certificates_at_risk = serializers.CharField(required=False, allow_blank=True, max_length=100)
    is_fleetwide_relevance = serializers.BooleanField(required=False, default=False)

    def validate_clauses(self, value):
        if not value:
            raise serializers.ValidationError("At least one clause reference is required.")
        primary_count = sum(1 for clause in value if clause.get("is_primary"))
        if primary_count != 1:
            raise serializers.ValidationError("Exactly one clause reference must be marked primary.")
        return value


class AuditFindingResponseSerializer(serializers.Serializer):
    def to_representation(self, instance):
        finding = instance.finding
        deficiency = instance.deficiency
        car = instance.car
        clauses = _finding_clauses_for_response(finding.id)
        return {
            "id": str(finding.id),
            "audit_detail_id": str(finding.audit_detail_id),
            "finding_type": finding.finding_type,
            "nc_category": finding.nc_category,
            "observation_category": finding.observation_category,
            "standard_code": finding.standard_code,
            "rule_book_type": finding.rule_book_type,
            "rule_clause_id": str(finding.rule_clause_id) if finding.rule_clause_id else None,
            "clause_ref_text": finding.clause_ref_text,
            "description": finding.description or deficiency.description,
            "objective_evidence": finding.objective_evidence,
            "priority": finding.priority,
            "certificates_at_risk": finding.certificates_at_risk,
            "is_fleetwide_relevance": finding.is_fleetwide_relevance,
            "linked_circular_id": str(finding.linked_circular_id) if finding.linked_circular_id else None,
            "psc_deficiency_id": str(deficiency.id),
            "car_id": str(car.id),
            "car_number": car.car_number,
            "car_status": car.status,
            "created": instance.created,
            "clauses": [
                {
                    "id": str(clause.id),
                    "rule_book_type": clause.rule_book_type,
                    "rule_clause_id": str(clause.rule_clause_id) if clause.rule_clause_id else None,
                    "clause_ref_text": clause.clause_ref_text,
                    "clause_subref_text": clause.clause_subref_text,
                    "is_primary": clause.is_primary,
                }
                for clause in clauses
            ],
        }


def _finding_clauses_for_response(finding_id):
    if connection.vendor == "microsoft":
        return list(
            AuditFindingClause.objects.raw(
                f"""
                SELECT *
                FROM dbo.{AuditFindingClause._meta.db_table}
                WHERE audit_finding_id = CAST(%s AS uniqueidentifier)
                  AND is_deleted = 0
                ORDER BY is_primary DESC, created_date ASC, id ASC
                """,
                [str(finding_id)],
            )
        )
    return AuditFindingClause.objects.filter(audit_finding_id=finding_id).order_by(
        "-is_primary",
        "created_date",
        "id",
    )
